### PR TITLE
Added a function to generate decoys from scrambled targets

### PR DESCRIPTION
**Pull request by zdanaceau**
This PR enables decoys to be created from targets, which can be done at run-time. Doing this at runtime allows us to make sure the decoy is sufficiently different from the target. It also allows us to have paired targets and decoys, which is advantages for FDR control. see https://www.biorxiv.org/content/10.1101/2022.05.11.491571v1
Eventually we will compare spectra, rather than sequence, to insure dissimilarity.

### CODE REVIEW

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (Alexander-Sol) [inline mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1402]: Function name is misspelled, length should be stored as an int, not a double

**Comment by zdanaceau** [comment]
Only merge this pull request after TargetDecoyPairing is merged

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (zdanaceau) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1335]: I set it to a list rather than an array since I need to be able to pop elements out from it.

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (zdanaceau) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1352]: No because, the newBase array, which tempNewBase is a copy of already has these filled.  It is only '0' in the evaporatingBase to mark it as a residue not to fill.

Comment (zdanaceau) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1352]: I sent you a message on slack showing the two arrays.

File: mzLib/Test/TestPeptideWithSetMods.cs
Comment (nbollis) [inline mzLib/Test/TestPeptideWithSetMods.cs:875]: Would you be able to add tests for more complicated cases than PEPTIDEK

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (zdanaceau) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs]: Fixed all of these

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (nbollis) [inline mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1429]: Make sure we test the case where the newBaseString != this.BaseSequence

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (nbollis) [inline mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1394]: Please add a test that covers what happens if the scrambleAttemt == maxScrambles

Comment (zdanaceau) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1394]: I fixed this with the AVLRRRKKRDEF which forces scrambleAttempt == maxScrambles since that peptide cannot be scrambled S.T. sequence homology is < 0.3

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (zdanaceau) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1429]: I fixed this with the AVLRRRKKRDEF which forces scrambleAttempt == maxScrambles since that peptide cannot be scrambled S.T. sequence homology is < 0.3

**Review by nbollis — APPROVED** [review]

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (Alexander-Sol) [inline mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1264]: Why is revisedAminoAcidOrder a parameter? I feel like it would be simpler to initialize an empty array inside the method. Do you anticipate the revisedAminoAcidOrder array being useful outside of testing?

**Review by Alexander-Sol — CHANGES_REQUESTED** [review]

File: mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs
Comment (zdanaceau) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1264]: I kept it as a parameter because it was a parameter in the original which generated decoys by just reversing the sequence.

Comment (Alexander-Sol) [reply mzLib/Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:1264]: Sounds good

**Review by Alexander-Sol — APPROVED** [review]

**Pull request merged**
